ffi/src/winscard/buf_alloc.rs

```diff
@@ -1,121 +1,127 @@
-        }
+    Ok(if p_buf.is_null() {
+        // If this value is NULL, we ignore the buffer length, writes the length of the buffer that
+        // would have been returned if this parameter had not been NULL, and returns a success code.
+        RequestedBufferType::Length
+    } else if
+    // SAFETY: The `pcb_buf` parameter cannot be null. We've checked for it above.
+    unsafe { *pcb_buf } == SCARD_AUTOALLOCATE {
+        // If the buffer length is specified as SCARD_AUTOALLOCATE, then data pointer is
+        // converted to a pointer to a byte pointer, and receives the address of a block of memory
+        // containing the attribute.
+        RequestedBufferType::Allocate
     } else {
-        if buff_to_copy_len > unsafe { *raw_buff_len } {
-            return Err(Error::new(
-                ErrorKind::InsufficientBuffer,
-                format!(
-                    "expected at least {} bytes but got {}.",
-                    buff_to_copy_len, *raw_buff_len
-                ),
-            ));
-        }
-        unsafe {
-            *raw_buff_len = buff_to_copy_len;
-            from_raw_parts_mut(raw_buff, buff_to_copy.len()).copy_from_slice(buff_to_copy);
-        }
+        // SAFETY: `p_buf` and `pcb_buf` parameters can't be null. We've checked for it above.
+        RequestedBufferType::Buff(unsafe { from_raw_parts_mut(p_buf, (*pcb_buf).try_into()?) })
+    })
+}
+
+/// This function behaves as the [build_buf_request_type] but here it expects a pointer
+/// to the `u16` buffer instead of `u8`. So, the buffer length is multiplied by two.
+pub unsafe fn build_buf_request_type_wide<'data>(
+    p_buf: LpWStr,
+    pcb_buf: LpDword,
+) -> WinScardResult<RequestedBufferType<'data>> {
+    if pcb_buf.is_null() {
+        return Err(Error::new(ErrorKind::InvalidParameter, "pcb_buf cannot be null"));
     }
 
-    Ok(())
+    Ok(if p_buf.is_null() {
+        // If this value is NULL, SCardGetAttrib ignores the buffer length supplied in pcbAttrLen,
+        // writes the length of the buffer that would have been returned if this parameter had not been NULL
+        // to pcbAttrLen, and returns a success code.
+        RequestedBufferType::Length
+    } else if
+    // SAFETY: The `pcb_buf` parameter cannot be null. We've checked for it above.
+    unsafe { *pcb_buf } == SCARD_AUTOALLOCATE {
+        // https://learn.microsoft.com/en-us/windows/win32/api/winscard/nf-winscard-scardgetattrib
+        //
+        // If the buffer length is specified as SCARD_AUTOALLOCATE, then pbAttr is converted to a pointer
+        // to a byte pointer, and receives the address of a block of memory containing the attribute.
+        RequestedBufferType::Allocate
+    } else {
+        // SAFETY: `p_buf` and `pcb_buf` parameters can't be null. We've checked for it above.
+        RequestedBufferType::Buff(unsafe { from_raw_parts_mut(p_buf as *mut u8, usize::try_from(*pcb_buf)? * 2) })
```

Comment:
thought: I think it was already like that since before, but the usual abbreviation is `Buf`. At some point we could rename that.